[MRG] Add screenshot method for GUI

chenghuzi opened the pull request:
We add a method `render_iframe` for the GUI object so you can use it to take a GUI snapshot. Notice that this operation creates static html files as it's using iframe to load the snapshot.

**Warning**: do not use the example tutorial notebook with JupyterLab, use `Jupyter-notebook`.

chenghuzi commented:
We may need to consider including html files in doc generation

jasmainak commented:
I just read the javascript code, seems neat! Can it be put in the `HNNGUI` class?

chenghuzi commented:
> I just read the javascript code, seems neat! Can it be put in the `HNNGUI` class?

I've put the long js string into the GUI class but to call it in notebooks it seems we still need the `Javascript` method in a notebook cell.

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+
+    def take_screenshot(self, width=None, height=None, filename=None,
+                        render=True):
+        """Embed the widget to HTML and render it through iframe.
```
```suggestion
        """Embed the widget to HTML and render it through iframe.

```

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+        """Embed the widget to HTML and render it through iframe.
+        Parameters
+        ----------
+        width : Optional[int]
```
```suggestion
        width : int
```

to be consistent with rest of repo

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+            The width of iframe window use to show the snapshot.
+        height : Optional[int]
+            The height of iframe window use to show the snapshot.
+        filename : Optional[int]
```
```suggestion
        filename : str | None
```

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+        ----------
+        width : Optional[int]
+            The width of iframe window use to show the snapshot.
+        height : Optional[int]
```
int

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+    def show(self):
+        display(self.app_layout)
+
+    def take_screenshot(self, width=None, height=None, filename=None,
```
```suggestion
    def show(self, freeze=True,
```

? Now that you have it working, it's helpful to think of the API a bit. Screenshot usually means it is not interactive ...

chenghuzi replied:
It's indeed not interactive, by which I mean it will not influence the data model in the kernel. Probably `capture` could be a better name?

chenghuzi replied:
Use `capture` now

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+
+    def run_notebook_cells(self):
+        """Run all but the last cells sequentially in a Jupyter notebook.
+        To properly use this function:
```
```suggestion

        To properly use this function:
```

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+        """
+        self._screenshots_count += 1
+        if not filename:
+            filename = f"snapshot_{self._screenshots_count}.html"
```
I would suggest saving to a file in the `/tmp/` directory using this library: https://docs.python.org/3/library/tempfile.html

Why do you need to keep track of the filename or screenshot count?

chenghuzi replied:
Because we load static html files from disk to show the "screenshot". It's actually an `<iframe>` object with local URL. And the user may take multiple screenshots at different time, so we need a mechanism to assign each screenshot a different name so that they won't interfere with each other.

jasmainak replied:
let's use `tempfile` ... I don't want to pollute the user's directory. First make a temp directory, then store all the screenshots within that directory with filenames generated by `tempfile`

jasmainak replied:
another idea, probably even cleaner:

```py
import io
from IPython.display import HTML

html_code = io.StringIO()
embed_minimal_html(html_code, views=[self.app_layout], title='')
HTML(html_code)
```

chenghuzi replied:
We now do not need static files anymore! With dataurl we can load the widget content in to the `src` attribute of the `iframe` object. See 88d67460d17cd53358c9e48779fb0bc8805bf5de

jasmainak replied:
nice!

jasmainak reviewed (commented):
Did you try using `nbconvert`? Does that work with your screenshot method?

Ultimately what I am imagining is that we will have a folder with the three notebooks and when we do:

```sh
$ make html
```

it will first build the HNN-core examples, and then the GUI examples.

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+        self._screenshots_count += 1
+        if not filename:
+            filename = f"snapshot_{self._screenshots_count}.html"
+        embed_minimal_html(filename, views=[self.app_layout], title='')
```
how come this works now? Was the trick to pass `self.app_layout` to `views` ?

chenghuzi replied:
Inside `embed_minimal_html`, there is a function that extract data models from the widget list it received. Then that function turns each data model into a json object so that it can be pickled and later loaded. The actually stored objects are not html contents but a list of json objects.

jasmainak replied:
so is that how you made it work? I remember this was not working on Friday ...

chenghuzi replied:
No on Friday the problem is more about the callback delay. Basically we bypassed that problem by sequentially execute all cells with JS

chenghuzi commented:
> 

Yes I tried nbconvert and it works! Now I'm testing the `make html`

jasmainak commented on examples/gui/.gitignore:
```diff
@@ -0,0 +1,2 @@
+*.html
```
don't we have a .gitignore at the root level? You should update that

chenghuzi replied:
No need for this since last few commits so I removed it

jasmainak commented:
Do you want to merge this PR and continue the `Makefile` edits in a new PR? I think it's better if you remove the notebook before we merge. We should do a separate PR for each tutorial ...

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -98,25 +104,31 @@ def __init__(self, theme_color="#8A2BE2", log_window_height="100px",
                  visualization_window_height="500px",
                  left_sidebar_width='380px',
-                 drive_widget_width="200px"):
+                 drive_widget_width="200px",
+                 header_height="50px",
+                 button_height="30px",
```
I think we shouldn't expose all these parameters ... the GUI should be responsive. Everything should be defined as percentages of the total width/height available. Maybe you can allow users to specify the vertical and horizontal split location ... but nothing else

chenghuzi replied:
I don't think that would be a good idea to make everything responsive ... Responsiveness is good and fancy indeed for web but it also causes problems as some parts in our GUI need a minimal size and the `Layout` in ipywidget is not so flexible that we can customize all CSS attribute. If users think the GUI is too large in the browser page, they should resize the browser.

jasmainak replied:
okay let's keep it like this for now ... at least having the layout in one place makes it easy to modify :)

chenghuzi commented:
> Did you try using `nbconvert`? Does that work with your screenshot method?
> 
> Ultimately what I am imagining is that we will have a folder with the three notebooks and when we do:
> 
> ```shell
> $ make html
> ```
> 
> it will first build the HNN-core examples, and then the GUI examples.

`make html` now works perfectly and it generates something like this:
![image](https://user-images.githubusercontent.com/11160442/182047043-f5b696f6-637a-498b-ba35-76e4b199d6bf.png)

Instead of using sphinx-gallary, I added[nbsphinx](https://nbsphinx.readthedocs.io/en/0.8.9/) to run the ipynb files and render the html. To do that I moved the tutorial notebook outside the `gui` folder and create a new folder under `doc`.
The reason not to use sphinx-gallary is that sphinx-gallary use python scripts to generate static files for the doc site and the resulting html file does not include the iframe output. Instead, it generate something pure-text like `AppLayout...`. So I gave up and started to use nbsphinx directly.

chenghuzi commented:
> 



> Do you want to merge this PR and continue the `Makefile` edits in a new PR? I think it's better if you remove the notebook before we merge. We should do a separate PR for each tutorial ...

Yes I'd like to merge it as the gui code part is finished.

jasmainak commented on doc/gui/index.rst:
```diff
@@ -0,0 +1,11 @@
+ GUI
+=====
+
+The bran-new light-weight web-based hnn GUI that works with hnn-core.
```
```suggestion
The brand-new lightweight web-based hnn GUI that works with hnn-core.
```

jasmainak commented:
Wonderful! Great job 🥳 🥳 🥳 

Now my only concern is that CIs will time-out, specially if we're running the notebooks twice ...

can you update the contributing.rst doc to indicate that developers have to install nbsphinx?

jasmainak commented on hnn_core/gui/gui.py:
```diff
@@ -398,7 +421,102 @@ def compose(self):
+        else:
+            return self.app_layout
+
+    def show(self):
+        display(self.app_layout)
```
```suggestion
```

jasmainak replied:
I guess it's not used anywhere

chenghuzi commented:
> Wonderful! Great job 🥳 🥳 🥳
> 
> Now my only concern is that CIs will time-out, specially if we're running the notebooks twice ...

I don't think we will hit the wall at this moment, but when we have more tutorials that could be the case... How about running those notebooks locally and then directly render the html? We can force the nbsphinx not to execute them. Anyway, I think it's not a big issue and we can fix that later.

> can you update the contributing.rst doc to indicate that developers have to install nbsphinx?

Sure!

jasmainak commented:
we could probably copy [what MNE-Python does](https://github.com/mne-tools/mne-python/blob/3b66a51132ba6de184ae3e1750cc16b7e2053baf/.circleci/config.yml#L1-L10) ... build only GUI examples that have changed or build an example only every third commit. But let's do it when we hit the wall ...

jasmainak commented:
I'll merge once the CIs come green ... I want to look at the built html

jasmainak commented:
@rythorpe any idea about the MPI test fail?

jasmainak commented:
okay looks like the tests pass.

Really nice: https://output.circle-artifacts.com/output/job/0ba01975-6a33-43e9-a44d-3608ec7eb63f/artifacts/0/html/gui/basic_gui_usage.html !

Let's merge.

@chenghuzi what are the next steps? Should we go ahead and write the evoked response tutorial?! I'm excited 🥳

jasmainak commented:
This will give you a headstart @chenghuzi :

https://github.com/jonescompneurolab/hnn-workshop-materials/blob/master/erp_tutorial_hnn_core.ipynb
https://jonescompneurolab.github.io/hnn-tutorials/erp/erp

rythorpe commented:
> @rythorpe any idea about the MPI test fail?

It looks like the problem was resolved @jasmainak?

jasmainak commented:
yeah ... this is an intermittent issue

rythorpe commented:
I'm assuming it was the age-old [`RuntimeError: MPI simulation failed. Return code: 143`](https://github.com/jonescompneurolab/hnn-core/issues/406)?

jasmainak commented:
yes!
